File: frontend/src/api/resources.ts
/**
 * API functions for resources (учет техники и людей) section
 */

import { request, API_VERSION } from './client'

const API_BASE = `/api/${API_VERSION}/resources`

// =============================================================================
// Types - Reference Data
// =============================================================================

export interface EquipmentType {
  id: number
  name: string
  is_active: boolean
}

export interface Vehicle {
  id: number
  equipment_type_id: number
  equipment_type_name: string | null
  plate_number: string
  is_active: boolean
}

export interface Driver {
  id: number
  full_name: string
  phone: string | null
  is_active: boolean
}

export interface Master {
  id: number
  full_name: string
  phone: string | null
  is_active: boolean
}

export interface EquipmentTypesResponse {
  items: EquipmentType[]
}

export interface VehiclesResponse {
  items: Vehicle[]
}

export interface DriversResponse {
  items: Driver[]
}

export interface MastersResponse {
  items: Master[]
}

// =============================================================================
// Types - Equipment Shifts
// =============================================================================

export interface EquipmentShiftCreate {
  is_own: boolean
  vehicle_id?: number | null
  driver_id?: number | null
  equipment_type_id: number
  plate_number: string
  driver_name?: string | null
  shift_start_date: string  // YYYY-MM-DD
  shift_start_time: string  // HH:MM or HH:MM:SS
  shift_duration_hours: number
}

export interface EquipmentShiftUpdate {
  equipment_type_id?: number | null
  driver_id?: number | null
  driver_name?: string | null
  shift_start_time?: string | null
  shift_duration_hours?: number | null
}

export interface EquipmentShiftResponse {
  id: number
  is_own: boolean
  vehicle_id: number | null
  equipment_type_id: number
  equipment_type_name: string | null
  plate_number: string
  driver_id: number | null
  driver_name: string | null
  shift_start_date: string
  shift_start_time: string
  shift_start_at: string | null
  shift_duration_hours: number
  created_at: string | null
  updated_at: string | null
  is_deleted: boolean
}

export interface EquipmentShiftSearchRequest {
  plate_number: string
  shift_start_date: string  // YYYY-MM-DD
}

export interface EquipmentShiftDeleteRequest {
  delete_reason: string
}

// =============================================================================
// Types - Master Shifts
// =============================================================================

export interface MasterShiftCreate {
  master_id: number
  workers_count: number
  shift_start_date: string  // YYYY-MM-DD
  shift_start_time: string  // HH:MM or HH:MM:SS
  shift_duration_hours: number
}

export interface MasterShiftUpdate {
  workers_count?: number | null
  shift_start_time?: string | null
  shift_duration_hours?: number | null
}

export interface MasterShiftResponse {
  id: number
  master_id: number
  master_full_name: string | null
  workers_count: number
  shift_start_date: string
  shift_start_time: string
  shift_start_at: string | null
  shift_duration_hours: number
  created_at: string | null
  updated_at: string | null
  is_deleted: boolean
}

export interface MasterShiftSearchRequest {
  master_id: number
  shift_start_date: string  // YYYY-MM-DD
}

export interface MasterShiftDeleteRequest {
  delete_reason: string
}

// =============================================================================
// Types - Summary
// =============================================================================

export interface SummaryEquipmentItem {
  equipment_type_id: number
  equipment_type_name: string
  count: number
}

export interface SummaryEquipmentGroup {
  is_own: boolean
  label: string
  items: SummaryEquipmentItem[]
  total: number
}

export interface SummaryEquipment {
  groups: SummaryEquipmentGroup[]
  grand_total: number
}

export interface SummaryMasterItem {
  master_id: number
  master_full_name: string
  workers_count: number
}

export interface SummaryPeople {
  masters: SummaryMasterItem[]
  total_masters: number
  total_workers: number
  grand_total: number
}

export interface SummaryResponse {
  date: string
  time_from: string | null
  time_to: string | null
  equipment: SummaryEquipment
  people: SummaryPeople
}

// =============================================================================
// Types - Generic
// =============================================================================

export interface MessageResponse {
  message: string
  success: boolean
}

// =============================================================================
// API Functions - Reference Data
// =============================================================================

/**
 * Get list of active equipment types
 */
export async function getEquipmentTypes(): Promise<EquipmentTypesResponse> {
  return await request<EquipmentTypesResponse>(`${API_BASE}/references/equipment-types`)
}

/**
 * Get list of active vehicles, optionally filtered by equipment type
 */
export async function getVehicles(equipmentTypeId?: number): Promise<VehiclesResponse> {
  const params = new URLSearchParams()
  if (equipmentTypeId !== undefined) {
    params.set('equipment_type_id', String(equipmentTypeId))
  }
  const queryString = params.toString()
  const url = queryString ? `${API_BASE}/references/vehicles?${queryString}` : `${API_BASE}/references/vehicles`
  return await request<VehiclesResponse>(url)
}

/**
 * Get list of active drivers
 */
export async function getDrivers(): Promise<DriversResponse> {
  return await request<DriversResponse>(`${API_BASE}/references/drivers`)
}

/**
 * Get list of active masters
 */
export async function getMasters(): Promise<MastersResponse> {
  return await request<MastersResponse>(`${API_BASE}/references/masters`)
}

// =============================================================================
// API Functions - Equipment Shifts
// =============================================================================

/**
 * Create a new equipment shift
 */
export async function createEquipmentShift(data: EquipmentShiftCreate): Promise<EquipmentShiftResponse> {
  return await request<EquipmentShiftResponse>(`${API_BASE}/equipment-shifts`, {
    method: 'POST',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Search for equipment shift by plate number and date
 */
export async function searchEquipmentShift(data: EquipmentShiftSearchRequest): Promise<EquipmentShiftResponse> {
  return await request<EquipmentShiftResponse>(`${API_BASE}/equipment-shifts/search`, {
    method: 'POST',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Update equipment shift
 */
export async function updateEquipmentShift(shiftId: number, data: EquipmentShiftUpdate): Promise<EquipmentShiftResponse> {
  return await request<EquipmentShiftResponse>(`${API_BASE}/equipment-shifts/${shiftId}`, {
    method: 'PUT',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Delete (soft) equipment shift
 */
export async function deleteEquipmentShift(shiftId: number, data: EquipmentShiftDeleteRequest): Promise<MessageResponse> {
  return await request<MessageResponse>(`${API_BASE}/equipment-shifts/${shiftId}`, {
    method: 'DELETE',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

// =============================================================================
// API Functions - Master Shifts
// =============================================================================

/**
 * Create a new master shift
 */
export async function createMasterShift(data: MasterShiftCreate): Promise<MasterShiftResponse> {
  return await request<MasterShiftResponse>(`${API_BASE}/master-shifts`, {
    method: 'POST',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Search for master shift by master_id and date
 */
export async function searchMasterShift(data: MasterShiftSearchRequest): Promise<MasterShiftResponse> {
  return await request<MasterShiftResponse>(`${API_BASE}/master-shifts/search`, {
    method: 'POST',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Update master shift
 */
export async function updateMasterShift(shiftId: number, data: MasterShiftUpdate): Promise<MasterShiftResponse> {
  return await request<MasterShiftResponse>(`${API_BASE}/master-shifts/${shiftId}`, {
    method: 'PUT',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

/**
 * Delete (soft) master shift
 */
export async function deleteMasterShift(shiftId: number, data: MasterShiftDeleteRequest): Promise<MessageResponse> {
  return await request<MessageResponse>(`${API_BASE}/master-shifts/${shiftId}`, {
    method: 'DELETE',
    body: data as unknown as BodyInit,
    credentials: 'include',
  })
}

// =============================================================================
// API Functions - Summary
// =============================================================================

/**
 * Get summary of equipment and people for a given date/time range
 */
export async function getSummary(params: {
  date: string  // YYYY-MM-DD
  time_from?: string  // HH:MM
  time_to?: string  // HH:MM
}): Promise<SummaryResponse> {
  const queryParams = new URLSearchParams()
  queryParams.set('date', params.date)
  if (params.time_from) queryParams.set('time_from', params.time_from)
  if (params.time_to) queryParams.set('time_to', params.time_to)
  
  return await request<SummaryResponse>(`${API_BASE}/summary?${queryParams.toString()}`)
}

File: app/backend/schemas/resources.py
"""Schemas for resources (учет техники и людей) section."""

from typing import List, Optional
from datetime import date, time, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class _BaseSchema(BaseModel):
    """Base schema with optimized serialization settings."""
    model_config = ConfigDict(
        extra="ignore",
        from_attributes=True,
        ser_json_inf_nan="constants",
    )


# =============================================================================
# Reference Data Schemas (Справочники)
# =============================================================================

class EquipmentType(_BaseSchema):
    """Equipment type reference."""
    id: int
    name: str
    is_active: bool = True


class Vehicle(_BaseSchema):
    """Own vehicle (собственная техника)."""
    id: int
    equipment_type_id: int
    plate_number: str
    is_active: bool = True


class VehicleWithType(_BaseSchema):
    """Vehicle with equipment type name for display."""
    id: int
    equipment_type_id: int
    equipment_type_name: Optional[str] = None
    plate_number: str
    is_active: bool = True


class Driver(_BaseSchema):
    """Driver reference."""
    id: int
    full_name: str
    phone: Optional[str] = None
    is_active: bool = True


class Master(_BaseSchema):
    """Master reference."""
    id: int
    full_name: str
    phone: Optional[str] = None
    is_active: bool = True


# =============================================================================
# Equipment Shift Schemas (Смены техники)
# =============================================================================

class EquipmentShiftCreate(_BaseSchema):
    """Request to create equipment shift."""
    is_own: bool
    # For own equipment
    vehicle_id: Optional[int] = None
    driver_id: Optional[int] = None
    # For both own and rented
    equipment_type_id: int
    plate_number: str
    driver_name: Optional[str] = None  # For rented equipment or as fallback
    # Shift details
    shift_start_date: date
    shift_start_time: time
    shift_duration_hours: Decimal = Field(..., ge=0.5, le=24)


class EquipmentShiftUpdate(_BaseSchema):
    """Request to update equipment shift (limited fields)."""
    # Can't change: plate_number, shift_start_date, is_own, vehicle_id
    equipment_type_id: Optional[int] = None
    driver_id: Optional[int] = None
    driver_name: Optional[str] = None
    shift_start_time: Optional[time] = None
    shift_duration_hours: Optional[Decimal] = Field(None, ge=0.5, le=24)


class EquipmentShiftResponse(_BaseSchema):
    """Equipment shift response."""
    id: int
    is_own: bool
    vehicle_id: Optional[int] = None
    equipment_type_id: int
    equipment_type_name: Optional[str] = None
    plate_number: str
    driver_id: Optional[int] = None
    driver_name: Optional[str] = None
    shift_start_date: date
    shift_start_time: time
    shift_start_at: Optional[datetime] = None
    shift_duration_hours: Decimal
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    is_deleted: bool = False


class EquipmentShiftSearchRequest(_BaseSchema):
    """Request to search for equipment shift by plate and date."""
    plate_number: str
    shift_start_date: date


class EquipmentShiftDeleteRequest(_BaseSchema):
    """Request to delete (soft) equipment shift."""
    delete_reason: str = Field(..., min_length=5, max_length=100)


# =============================================================================
# Master Shift Schemas (Смены мастеров)
# =============================================================================

class MasterShiftCreate(_BaseSchema):
    """Request to create master shift."""
    master_id: int
    workers_count: int = Field(..., ge=0)
    shift_start_date: date
    shift_start_time: time
    shift_duration_hours: Decimal = Field(..., ge=0.5, le=24)


class MasterShiftUpdate(_BaseSchema):
    """Request to update master shift (limited fields)."""
    # Can't change: master_id, shift_start_date
    workers_count: Optional[int] = Field(None, ge=0)
    shift_start_time: Optional[time] = None
    shift_duration_hours: Optional[Decimal] = Field(None, ge=0.5, le=24)


class MasterShiftResponse(_BaseSchema):
    """Master shift response."""
    id: int
    master_id: int
    master_full_name: Optional[str] = None
    workers_count: int
    shift_start_date: date
    shift_start_time: time
    shift_start_at: Optional[datetime] = None
    shift_duration_hours: Decimal
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    is_deleted: bool = False


class MasterShiftSearchRequest(_BaseSchema):
    """Request to search for master shift by master_id and date."""
    master_id: int
    shift_start_date: date


class MasterShiftDeleteRequest(_BaseSchema):
    """Request to delete (soft) master shift."""
    delete_reason: str = Field(..., min_length=5, max_length=100)


# =============================================================================
# Summary Schemas (Сводка)
# =============================================================================

class SummaryEquipmentItem(_BaseSchema):
    """Single equipment item in summary."""
    equipment_type_id: int
    equipment_type_name: str
    count: int


class SummaryEquipmentGroup(_BaseSchema):
    """Group of equipment (own or rented) in summary."""
    is_own: bool
    label: str  # "Собственная" or "Арендованная"
    items: List[SummaryEquipmentItem]
    total: int


class SummaryEquipment(_BaseSchema):
    """Equipment summary section."""
    groups: List[SummaryEquipmentGroup]
    grand_total: int


class SummaryMasterItem(_BaseSchema):
    """Single master in summary."""
    master_id: int
    master_full_name: str
    workers_count: int


class SummaryPeople(_BaseSchema):
    """People (masters + workers) summary section."""
    masters: List[SummaryMasterItem]
    total_masters: int
    total_workers: int
    grand_total: int  # masters + workers


class SummaryResponse(_BaseSchema):
    """Full summary response."""
    date: date
    time_from: Optional[time] = None
    time_to: Optional[time] = None
    equipment: SummaryEquipment
    people: SummaryPeople


# =============================================================================
# Reference List Responses
# =============================================================================

class EquipmentTypesResponse(_BaseSchema):
    """List of equipment types."""
    items: List[EquipmentType]


class VehiclesResponse(_BaseSchema):
    """List of vehicles."""
    items: List[VehicleWithType]


class DriversResponse(_BaseSchema):
    """List of drivers."""
    items: List[Driver]


class MastersResponse(_BaseSchema):
    """List of masters."""
    items: List[Master]


class RentedPlateNumber(_BaseSchema):
    """Rented equipment plate number."""
    plate_number: str
    equipment_type_id: int
    equipment_type_name: Optional[str] = None


class RentedPlateNumbersResponse(_BaseSchema):
    """List of unique rented plate numbers."""
    items: List[RentedPlateNumber]


# =============================================================================
# Generic Response Wrappers
# =============================================================================

class MessageResponse(_BaseSchema):
    """Generic message response."""
    message: str
    success: bool = True
